src/persist-client/src/internal/compact.rs

```diff
@@ -184,82 +192,416 @@ impl Compactor {
-            // to do this has already been established in old persist (see
-            // background.rs). We'll have to generalize that from 2 to N batches
-            // and also resolve the (much much harder) issue of new persist
-            // batch sorted-ness.
-            let mut updates = Vec::new();
-            for batch in req.inputs.iter() {
-                for part in batch.parts.iter() {
-                    let mut fetched_part = fetch_batch_part(
-                        &req.shard_id,
-                        blob.as_ref(),
-                        &metrics,
-                        &part.key,
-                        &batch.desc,
-                    )
-                    .await?;
-                    while let Some((k, v, mut t, d)) = fetched_part.next() {
-                        t.advance_by(req.desc.since().borrow());
+            let (parts, runs, updates) = Self::compact_runs::<T, D>(
+                &cfg,
+                &req.shard_id,
+                &req.desc,
+                runs,
+                batch_parts,
+                Arc::clone(&blob),
+                Arc::clone(&metrics),
+            )
+            .await?;
+            assert!((updates == 0 && parts.len() == 0) || (updates > 0 && parts.len() > 0));
+
+            if updates == 0 {
+                continue;
+            }
+            // merge together parts and runs from each compaction round.
+            // parts are appended onto our existing vec, and then we shift
+            // the latest run offsets to account for prior parts.
+            //
+            // e.g. if we currently have 3 parts and 2 runs (including the implicit one from 0):
+            //         parts: [k0, k1, k2]
+            //         runs:  [    1     ]
+            //
+            // and we merge in another result with 2 parts and 2 runs:
+            //         parts: [k3, k4]
+            //         runs:  [    1]
+            //
+            // we our result will contain 5 parts and 4 runs:
+            //         parts: [k0, k1, k2, k3, k4]
+            //         runs:  [    1       3   4 ]
+            let run_offset = all_parts.len();
+            if all_parts.len() > 0 {
+                all_runs.push(run_offset);
+            }
+            all_runs.extend(runs.iter().map(|run_start| run_start + run_offset));
+            all_parts.extend(parts);
+            len += updates;
+        }
+
+        Ok(CompactRes {
+            output: HollowBatch {
+                desc: req.desc.clone(),
+                parts: all_parts,
+                runs: all_runs,
+                len,
+            },
+        })
+    }
+
+    /// Sorts and groups all runs from the inputs into chunks, each of which has been determined
+    /// to consume no more than `run_reserved_memory_bytes` at a time. Uses [Self::order_runs] to
+    /// determine the order in which runs are selected.
+    fn chunk_runs<'a, T, D>(
```

Comment:
I think we need to ensure this always selects at least two runs, even if that requires it to violate the memory requirements (e.g. maybe the inputs got written with a different, higher part size limit or something) so that it's easier to convince ourselves that things eventually converge on 1 run

Reply:
hm trying to think this through -- should the goal be that every chunk is >=2 runs (where possible, could have an odd number of runs and 1 gets left out), or that, for all chunks, there exists at least 1 with >=2 runs? the latter case seems like it still converges, albeit slowly, and could let us minimize the memory overage

Reply:
oh good q. the former makes more sense to me. if we're going to blip memory usage above the bound, may as well do it several time and converse faster

Reply:
I think in practice it still works to skip this, because if 2 runs are too large to fit in memory, then each one will be individually compacted and written back out into part sizes that _can_ fit 2 in memory at a time. that said.... I like the insurance to force it to be true, so I added in the logic + a metric we can check to see how often it fires / if it matches our expectations

Reply:
:ok_hand: